flink-connector-prometheus/src/main/java/org/apache/flink/connector/prometheus/table/PrometheusDynamicSink.java

```diff
@@ -0,0 +1,193 @@
+                maxInFlightRequests,
+                maxBufferedRequests,
+                maxBufferSizeInBytes, // maxRecordSizeInSamples
+                maxTimeInBufferMS);
+
+        this.physicalDataType =
+                Preconditions.checkNotNull(physicalDataType, "Consumed data type must not be null");
+        this.prometheusConfig = prometheusConfig;
+    }
+
+    @Override
+    public ChangelogMode getChangelogMode(ChangelogMode changelogMode) {
+        return ChangelogMode.upsert();
+    }
+
+    @Override
+    public SinkRuntimeProvider getSinkRuntimeProvider(Context context) {
+
+        PrometheusSinkBuilder<RowData> builder =
+                new PrometheusSinkBuilder<RowData>()
+                        .setRetryConfiguration(
+                                PrometheusSinkConfiguration.RetryConfiguration
+                                        .DEFAULT_RETRY_CONFIGURATION)
+                        .setElementConverter(
+                                new RowDataElementConverter(physicalDataType, prometheusConfig))
+                        .setPrometheusRemoteWriteUrl(prometheusConfig.getRemoteWriteEndpointUrl());
+
+        Optional.ofNullable(prometheusConfig.getRequestSignerIdentifier())
+                .ifPresent(region -> builder.setRequestSigner(getRequestSigner(prometheusConfig)));
+        Optional.ofNullable(maxBatchSize).ifPresent(builder::setMaxBatchSizeInSamples);
+        Optional.ofNullable(maxBufferSizeInBytes).ifPresent(builder::setMaxRecordSizeInSamples);
+
+        PrometheusSink<RowData> prometheusSink = builder.build();
+        return SinkV2Provider.of(prometheusSink);
+    }
+
+    @Override
+    public DynamicTableSink copy() {
+        return new PrometheusDynamicSink(
+                maxBatchSize,
+                maxInFlightRequests,
+                maxBufferedRequests,
+                maxBufferSizeInBytes,
+                maxTimeInBufferMS,
+                physicalDataType,
+                prometheusConfig);
+    }
+
+    @Override
+    public String asSummaryString() {
+        return "Prometheus";
+    }
+
+    @Override
+    public void applyStaticPartition(Map<String, String> partitions) {
+        // We don't need to do anything here because the Prometheus sink handles a static partition
+        // just like a normal partition.
+    }
+
+    private PrometheusRequestSigner getRequestSigner(PrometheusConfig prometheusConfig) {
+        ServiceLoader<PrometheusDynamicRequestSignerFactory> loader =
+                ServiceLoader.load(PrometheusDynamicRequestSignerFactory.class);
+        Iterator<PrometheusDynamicRequestSignerFactory> factories = loader.iterator();
+        String requestSignerIdentifier = prometheusConfig.getRequestSignerIdentifier();
+
+        while (true) {
+            try {
+                if (!factories.hasNext()) {
+                    break;
+                }
+
+                PrometheusDynamicRequestSignerFactory factory = factories.next();
```

Comment:
Would this be simpler as a for loop as well as incorporating line 129?

```
for(Iterator<PrometheusDynamicRequestSignerFactory> factories = loader.iterator(); factories.hasNext(); ) {
   PrometheusDynamicRequestSignerFactory factory = factories.next();
```

Reply:
I have converted it to a Stream to make this more readable